sw/host/opentitanlib/src/util/usr_access.rs

```diff
@@ -35,3 +66,34 @@ pub fn usr_access_get(bitstream: &[u8]) -> Result<u32> {
     Ok(usr_access)
 }
 
+/// Returns a 32-bit timestamp suitable to be used as a USR_ACCESS value:
+///
+/// |-------+-------+-------+-------------------------+-------+--------+--------|
+/// | Bits: | 31:27 | 26:23 | 22:17                   | 16:12 | 11:6   | 5:0    |
+/// |-------+-------+-------+-------------------------+-------+--------+--------|
+/// | Data: | Day   | Month | Year (since 2000, 0-63) | Hour  | Minute | Second |
+/// |-------+-------+-------+-------------------------+-------+--------+--------|
+///
+/// From
+/// https://www.xilinx.com/content/dam/xilinx/support/documents/application_notes/xapp1232-bitstream-id-with-usr_access.pdf
+pub fn usr_access_timestamp() -> u32 {
+    let now = Utc::now();
+    now.day() << 27
+        | now.month() << 23
+        | now.year_ce().1.checked_sub(2000u32).unwrap() << 17
```

Comment:
It looks like we're throwing away the year's sign bit. From my time on Chrome, I'm wary of making assumptions about clocks :)

At first, I thought this would be better: `(now.year() as u32).checked_sub(2000u32).unwrap() << 17`? But I guess Rust's will happily convert a negative i32 into a u32 with the same bit pattern.

I guess I don't care too strongly about this year math — the stakes are pretty low, right?

Reply:
We are not really throwing away the sign bit because `checked_sub()` returns `u32`, that's what it's checking.
AFAICT there is no risk of wrong answer here and this line should work as long as year >= 2000.

Reply:
I was imagining that `now.year() : i32` is negative, so [`now.year_ce()`](https://docs.rs/chrono/latest/src/chrono/traits.rs.html#11-18) returns `(false, _)`. When we say `now.year_ce().1`, that's the throwing away of the sign bit that I was referring to. For instance, if `now.year()` returns -2021, the current implementation `now.year_ce().1` would think it's 2022.

All that to say, I'm happy to assume that the system clock is correctly configured with a year >= 2000, but it would be great to document the assumption in a comment. Or better yet, `assert!(now.year() > 2022)` after line 80!

Reply:
I was imagining that `now.year() : i32` is negative, so [`now.year_ce()`](https://docs.rs/chrono/latest/src/chrono/traits.rs.html#11-18) returns `(false, _)`. When we say `now.year_ce().1`, that's the throwing away of the sign bit that I was referring to. For instance, if `now.year()` returns -2021, the current implementation `now.year_ce().1` would think it's 2022.

All that to say, I'm happy to assume that the system clock is correctly configured with a year >= 2000, but it would be great to document the assumption in a comment. Or better yet, `assert!(now.year() > 2022)` after line 80!

Reply:
Ah, I see what you mean. I think `checked_sub()` panic if year < 2000 is roughly the same as the assert, right?

Reply:
I think if the system says the year is -2021 (not possible in Unix time?), `now.year_ce().1` would evaluate to +2022, so the `checked_sub(2000u32)` is too late in the pipeline to panic in this scenario.

I think the `checked_sub` here is equivalent to asserting that `now.year() >= 2000 && now.year() > -2000` (might be off by one).

Reply:
So my intent with the explicit `assert!(now.year() >= 2000i32)` was just to rule out the negative case.